sql-tuning-best-practice.md

```diff
@@ -0,0 +1,751 @@
+- Optimizer Fundamentals
+- Statistics Management
+- How TiDB build A Execution Plan
+- Understand Execution Plan
+- Index Strategy in TiDB
+  - SQL Tuning with a Covered Index
+  - SQL Tuning with a Composite Index Involing Sorting
+  - SQL Tuning with a Composite Index for Efficient Filtering and Sorting
+  - Composite Index Strategy Guidelines
+  - The Cost of Indexing
+
+## Query Processing Workflow
+
+The client sends a SQL statement to the protocol layer of TiDB server. The protocol layer is responsible for handling the connection between TiDB Server and the client, receiving SQL statement from the client, and returning data to the client.
+
+To the right of the protocol layer is the Optimizer layer of TiDB Server, which is responsible for processing SQL statements. The process is as follows:
+
+1. SQL statement arrives at the SQL optimizer through the protocol layer and is first parsed into an Abstract Syntax Tree (AST).
+2. Pre-Process is primarily for Point Get. If it is a Point Get, the following-up optimization processes can be skipped, the next step jumps to the SQL Executor.
+3. After confirming that it is not a Point Get, the AST goes to the Logical Transformation, which rewrites the SQL logically based on certain rules.
+4. The AST that has gone through Logical Transformation will undergo Cost-Based Optimization.
+5. During Cost-Based Optimization, the optimizer considers statistics to determine how to select specific operators and finally generates a physically executable physical execution plan.
+6. The generated physical execution plan is sent to the SQL Executor of the TiDB database for execution.
+7. Unlike traditional databases, TiDB databases need to push down the execution plan to different TiKV for reading and writing data.
+
+![workflow](/media/sql-tuning/workflow.png)
+
+## Optimizer Fundamentals
+
+TiDB uses a cost-based optimizer (CBO) to determine the most efficient execution plan for a SQL statement. This optimizer evaluates different execution strategies and chooses the one with the lowest estimated cost. The cost is influenced by factors such as:
+
+- SQL
+- Schema Design
+- Statistics
+  - Table
+  - Index
+  - Column
+
+Based on the input, The cost model will produce the execution plan, which includes the details how the system execute the sql, including 
+- Access Method
+- Join Method
+- Join Order
+
+The optimizer is as good as the information it receives. Therefore, ensuring up-to-date statistics and well-designed indexes is critical.
+
+## Statistics Management
+
+Statistics are essential to the TiDB optimizer. TiDB uses statistics as input to the optimizer to estimate the number of rows processed in each plan step for a SQL statement.
+
+statistics is generally divided into two levels: table level and column level. 
+- For table-level statistics, it includes the total number of rows in the table and the number of rows that have been modified since the last collection of statistics. 
```

Comment:
```suggestion

- For table-level statistics, it includes the total number of rows in the table and the number of rows that have been modified since the last collection of statistics. 
```

```
sql-tuning-best-practice.md: 172: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- For table-level statistics, ..."]
```